apps/src/gamelab/spritelab/spriteUtils.js

```diff
@@ -0,0 +1,222 @@
+var behaviors = [];
+
+export var background;
+export var title = '';
+export var subtitle = '';
+
+export function reset() {
+  spriteId = 0;
+  nativeSpriteMap = {};
+  inputEvents = [];
+  behaviors = [];
+  background = 'white';
+  title = subtitle = '';
+}
+
+/**
+ * Returns a list of all sprites that have the specified animation.
+ * Called on each tick of the draw loop because animations can change throughout runtime.
+ * @param {string} animation - animation name
+ */
+function allSpritesWithAnimation(animation) {
+  let group = [];
+  Object.keys(nativeSpriteMap).forEach(spriteId => {
+    if (nativeSpriteMap[spriteId].getAnimationLabel() === animation) {
+      let sprite = nativeSpriteMap[spriteId];
+      if (sprite) {
+        group.push(sprite);
+      }
+    }
+  });
+  return group;
+}
+
+/**
+ * Returns a list of sprites, specified either by id or animation name.
+ * @param {(string|number)} spriteOrGroup - Either the id or the animation name
+ * @return {[Sprite]} List of sprites that match the parameter. Either a list containing the one sprite
+ * the specified id, or a list containing all sprites with the specified animation.
+ */
+export function getSpriteArray(spriteOrGroup) {
+  if (typeof spriteOrGroup === 'number') {
+    const sprite = nativeSpriteMap[spriteOrGroup];
+    if (sprite) {
+      return [sprite];
+    }
+  }
+  if (typeof spriteOrGroup === 'string') {
+    return allSpritesWithAnimation(spriteOrGroup);
+  }
+  return [];
+}
+
+export function getAnimationsInUse() {
+  let animations = new Set();
+  Object.keys(nativeSpriteMap).forEach(spriteId => {
+    animations.add(nativeSpriteMap[spriteId].getAnimationLabel());
+  });
+  return Array.from(animations);
+}
+
+/**
+ * @param {string} animation
+ * @return {number} Number of behaviors associated with the specified animation.
+ */
+export function getBehaviorsForAnimation(animation) {
+  let numBehaviors = 0;
+  behaviors.forEach(behavior => {
+    if (behavior.sprite.getAnimationLabel() === animation) {
+      numBehaviors++;
+    }
+  });
+  return numBehaviors;
+}
+
+export function getSpriteIdsInUse() {
+  let spriteIds = [];
+  Object.keys(nativeSpriteMap).forEach(spriteId =>
+    spriteIds.push(parseInt(spriteId))
+  );
+  return spriteIds;
+}
+
+/**
+ * Adds the specified sprite to the native sprite map
+ * @param {Sprite} sprite
+ * @returns {Number} A unique id to reference the sprite.
+ */
+export function addSprite(sprite) {
+  nativeSpriteMap[spriteId] = sprite;
+  sprite.id = spriteId;
+  spriteId++;
+  return sprite.id;
+}
+
+/**
+ * Removes a sprite from the native sprite map
+ * @param {Number} spriteId
+ */
+export function deleteSprite(spriteId) {
+  delete nativeSpriteMap[spriteId];
+}
+
+export function addEvent(type, args, callback) {
+  inputEvents.push({type: type, args: args, callback: callback});
+}
+
+function checkEvent(inputEvent, p5Inst) {
+  let shouldEventFire = false;
+  let extraArgs = {};
+  switch (inputEvent.type) {
```

Comment:
Might just be my personal preference here: I tend to find big switch blocks with lots of inline logic a bit difficult to read. I usually prefer moving each of the case blocks to a function so the switch block itself is much more compact. i.e.
```
switch (type) {
  case 'whenpress': return onWhenPress(inputEvent, p5Inst);
...
```

Reply:
💯